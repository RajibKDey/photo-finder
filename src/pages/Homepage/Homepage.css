.homepage {
  text-align: center;
}

.param-div {
  position: sticky;
  top: 0px;
  background-color: #1c1f26;
}

.search {
  width: 400px;
  border-radius: 5px;
  padding: 10px;
  margin: 10px 0px;
}

.photo-list {
  display: flex;
  flex-wrap: wrap;
}

.no-more-results {
  background-color: grey;
  padding: 5px 0px;
  border-radius: 7px;
  text-align: center;
}

.no-results {
  display: flex;
  height: 90vh;
  justify-content: center;
  align-items: center;
}

.no-of-results {
  display: flex;
  justify-content: center;
  align-items: center;
}

.result-count {
  color: lightgray;
}

.no-result-img {
  padding: 20px 40px;
  border: 3px dotted grey;
}

.loading {
  display: flex;
  justify-content: center;
  padding: 20px;
}

.loader {
  border: 8px solid #f3f3f3; /* Light grey */
  border-top: 8px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 60px;
  height: 60px;
  animation: spin 2s linear infinite;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

@media only screen and (max-width: 600px) {
  .search {
    width: 100%;
    max-width: 400px;
  }
}
