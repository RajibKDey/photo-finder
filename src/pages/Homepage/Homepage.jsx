import React, { Component } from "react";

import { fetchPhotos } from "../../services/fetchPhotos";
import PhotoView from "../../components/PhotoView/PhotoView";
import noResult from '../../assets/images/no_result.png';

import './Homepage.css'

let timeout;
const debounce = (func, delay) => {
    clearTimeout(timeout);
    timeout = setTimeout(func, delay);
};

export default class Homepage extends Component {
    constructor() {
        super();
        this.state = {
            search: '',
            page: 1,
            pages: 0,
            photoList: [],
            loading: false,
        }
        this.handleChange = this.handleChange.bind(this);   
    }

    componentDidMount() {
        const { page } = this.state;
        document.addEventListener('scroll', this.trackScrolling);
        this.setState({ loading: true });
        fetchPhotos({ page }).then(res => {
            if(res){
                this.setState({photoList: res.photos.photo, pages: res.photos.pages, loading: false});
            }
        })
    }

    componentDidUpdate(_prevProps, prevState) {
        const {search: prevSearch} = prevState
        const {search} = this.state;
        if(search !== prevSearch){
            debounce(() => {
                this.setState({photoList: [], loading: true});
                fetchPhotos({search, page: 1}).then(res => {
                    if(res){
                        this.setState({photoList: res.photos.photo, pages: res.photos.pages, loading: false});
                    } else {
                        this.setState({photoList: [], loading: false});
                    }
                })
            }, 500)
        }
    }

    componentWillUnmount() {
        document.removeEventListener('scroll', this.trackScrolling);
    }

    isBottom(el) {
        return el.getBoundingClientRect().bottom <= window.innerHeight;
    }

    trackScrolling = () => {
        const { photoList, page, search, pages, loading } = this.state;
        const wrappedElement = document.getElementById('homepage');
        if (this.isBottom(wrappedElement)) {
            this.setState({ loading: true });
            if(!loading) {
                fetchPhotos({ page: page + 1, search }).then(res => {
                    if(res){
                        const tempPhotoList = [...photoList, ...res.photos.photo];
                        this.setState({photoList: tempPhotoList, pages: res.photos.pages, loading: false, page: res.photos.page});
                    }
                });
                console.log('homepage bottom reached');
                document.removeEventListener('scroll', this.trackScrolling);
            }
            if(page !== pages) {
                document.addEventListener('scroll', this.trackScrolling);
            }
        }
    };

    handleChange(e) {
        this.setState({search: e.target.value, page: 1, pages: 0});
    }

    render() {
        const { photoList, pages, page, loading, search } = this.state;
        return <div className='homepage' id='homepage'>
            <div className='param-div'>
                <input 
                    className='search'
                    id='search'
                    name='search'
                    type='text'
                    placeholder='Search'
                    label='Search'
                    onChange={this.handleChange}
                    value={search}
                />
            </div>
            {
                search && !loading && pages > 0 && <div className='no-of-results'>
                    <p className='result-count'>More than {pages*20} results found</p>
                </div>
            }
            <div className='photo-list'>
                {photoList.map(photo => <PhotoView key={photo.id} photo={photo} />)}
            </div>
            {
                loading && <div className='loading'>
                    <div className='loader'></div>
                </div>
            }
            {
                !loading && pages === 0 && <div className='no-results'>
                    <img className='no-result-img' src={noResult} alt='no results' />
                </div>
            }
            {
                pages > 0 && page > pages && <div className='no-more-results'>
                    <p>No more results to show</p>
                </div>
            }
            
        </div>
    }
}