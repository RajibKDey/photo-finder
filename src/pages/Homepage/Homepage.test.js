import React from 'react';
import { shallow } from 'enzyme';

import Homepage from './Homepage';

describe('Testing the Homepage Component', () => {
    let wrapper;
    beforeAll(() => {
        wrapper = shallow(<Homepage />);
    })

    afterEach(() => jest.clearAllMocks());

    it('should render the component without crashing', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should allow search', () => {
        wrapper.find('.search').hostNodes().simulate('change', {target: { value: 'search'}});
        expect(wrapper.state('search')).toEqual('search');
    });
});