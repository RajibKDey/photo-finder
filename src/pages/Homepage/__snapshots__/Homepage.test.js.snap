// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Homepage Component should render the component without crashing 1`] = `ShallowWrapper {}`;
